live-build/misc/upgrade-scripts/upgrade-container

```diff
@@ -164,6 +143,70 @@ function create() {
 	echo "$CONTAINER"
 }
 
+function create_upgrade_container_in_place() {
+	trap create_cleanup EXIT
+
+	#
+	# We want to use "/var/lib/machines" as the parent directory,
+	# since that directory is used by "systemd-nspawn" when looking
+	# for containers.  This way, we can create the directory and
+	# mount our rootfs clone here, and "systemd-nspawn" will
+	# automatically detect it.
+	#
+	DIRECTORY=$(mktemp -d -p "/var/lib/machines" -t in-place.XXXXXXX)
+	[[ -d "$DIRECTORY" ]] || die "failed to create upgrade directory"
+	CONTAINER=$(basename "$DIRECTORY")
+	[[ -n "$CONTAINER" ]] || die "failed to obtain upgrade name"
+
+	zfs snapshot "rpool/ROOT/ubuntu@$CONTAINER" ||
+		die "failed to create upgrade snapshot"
+	zfs clone -o mountpoint="$DIRECTORY" \
+		"rpool/ROOT/ubuntu@$CONTAINER" \
+		"rpool/ROOT/$CONTAINER" ||
+		die "failed to create upgrade clone"
+
+	create_upgrade_container_common
+}
+
+function create_upgrade_container_not_in_place() {
+	trap create_cleanup EXIT
+
+	#
+	# We want to use "/var/lib/machines" as the parent directory,
+	# since that directory is used by "systemd-nspawn" when looking
+	# for containers.  This way, we can create the directory and
+	# mount our new rootfs here, and "systemd-nspawn" will
+	# automatically detect it.
+	#
+	DIRECTORY=$(mktemp -d -p "/var/lib/machines" -t not-in-place.XXXXXXX)
+	[[ -d "$DIRECTORY" ]] || die "failed to create upgrade directory"
+	CONTAINER=$(basename "$DIRECTORY")
+	[[ -n "$CONTAINER" ]] || die "failed to obtain upgrade name"
+
+	zfs create -o mountpoint="$DIRECTORY" \
+		"rpool/ROOT/$CONTAINER" ||
+		die "failed to create upgrade filesystem"
+
+	#
+	# This function needs to return the container's name to stdout,
+	# so that consumers of this function/script can consume that
+	# name and then later start/stop/destroy the container. Thus, we
+	# have to redirect the output from debootstrap away from stdout.
+	#
+	# Also, we need to include the "systemd-container" package when
+	# installing the base systemd with debootstrap so that starting
+	# the container will work properly. Otherwise, after starting
+	# the container, we won't be able to communicate and later run
+	# commands in the container with "systemd-run".
+	#
+	debootstrap --no-check-gpg \
```

Comment:
Have you looked at `--variant`? Do we want to use `minbase`?

Reply:
I did, but now I don't recall the details. I could be mistaken, but I think the problem with that, is it doesn't create a system that can be booted with `systemd-nspawn`.

So, we could do that, then `chroot` in and install `delphix-entire`, without using `systemd-nspawn`, but I ran into issues where `delphix-entire` would fail to install in the `chroot` (i.e. dependent packages would fail to install).. this could have been due to not having things like `/sys`, `/proc`, etc. bind mounted, but I didn't try to root case that.

I opted for this approach since it integrates well into our existing code for creating the verification container using `systemd-nspawn`.